File: decl-py/tests/internal/session_test.py
"""session (tests/internal/checks.json): the operation log — apply, undo,
redo, and a new operation after undo discarding the redo tail."""

from __future__ import annotations

import json
from pathlib import Path

import pytest

from decl.semantics import path_str
from decl.session import Session, SessionError

ROOT = Path(__file__).resolve().parents[3]


def test_undo_redo() -> None:
    s = Session(str(ROOT / "tests/repl/documents/main.decl"))

    def bind(text: str) -> None:
        s.apply({"op": "bind", "name": "extra", "src": {"kind": "inline", "text": text}})

    bind('{ "port": 1, "name": "x" }')
    bound = s.document_text("extra")
    assert bound == '{"port":1,"name":"x"}'
    assert s.undo() == 1
    with pytest.raises(SessionError):  # undone: the root is invalid
        s.document_text("extra")
    assert s.redo() == 1
    assert s.document_text("extra") == bound
    s.undo()
    bind('{ "port": 2, "name": "y" }')
    assert s.redo() == 0, "a new operation after undo discards the redo tail"
    assert s.document_text("extra") == '{"port":2,"name":"y"}'


@pytest.mark.parametrize(
    "case", json.loads((ROOT / "tests/internal/edits.json").read_text()), ids=lambda c: c["name"]
)
def test_retained_edits(case: dict) -> None:
    entry = str(ROOT / "tests/internal/session_case.decl")
    overlay = {entry: case["source"]}
    retained, fresh = Session(entry, overlay), Session(entry, overlay)

    def apply(session: Session, op: dict, full: bool) -> None:
        Session.full_recompute = full
        try:
            if op["op"] == "undo":
                session.undo()
            elif op["op"] == "redo":
                session.redo()
            else:
                session.apply(op)
        finally:
            Session.full_recompute = False

    def snapshot(session: Session, full: bool):
        Session.full_recompute = full
        try:
            run = session.run()
            values = (
                [(n, run.eng.serialize(v, n)) for n, v in run.entry.env.roots.items()]
                if run.entry and run.eng
                else []
            )
            diags = [{k: v for k, v in d.items() if k != "by"} for d in run.diags]
            return run, (values, diags, run.load_diags, run.checks, run.session_checks)
        finally:
            Session.full_recompute = False

    for op in case["initial"]:
        apply(retained, op, False)
        apply(fresh, op, True)
    prior, actual = snapshot(retained, False)
    assert prior.eng is not None
    assert actual == snapshot(fresh, True)[1]
    for step in case["steps"]:
        assert prior.eng is not None and prior.entry is not None
        programs = prior.eng.programs
        inst = next(
            (r for r in prior.entry.env.registry if path_str(r.path) == step.get("retained")), None
        )
        edits = prior.eng.edits
        assert edits is not None
        verified, equal = edits.revisions.verified_cutoffs, edits.revisions.value_cutoffs
        apply(retained, step["action"], False)
        apply(fresh, step["action"], True)
        current, actual = snapshot(retained, False)
        expected = snapshot(fresh, True)[1]
        assert actual == expected, step
        assert programs is not None and current.eng is not None and programs is current.eng.programs
        if "retained" in step:
            assert inst is not None and inst in current.entry.env.registry
        if "max_prepared" in step:
            assert edits.prepared_queries <= step["max_prepared"]
        if "max_computes" in step:
            assert current.timing["recomputed"] <= step["max_computes"]
        if "min_verified" in step:
            assert edits.revisions.verified_cutoffs - verified >= step["min_verified"]
        if "min_value_cutoffs" in step:
            assert edits.revisions.value_cutoffs - equal >= step["min_value_cutoffs"]
        prior = current


def test_temporary_programs() -> None:
    case = json.loads((ROOT / "tests/internal/temporary.json").read_text())
    entry = str(ROOT / "tests/internal/temporary_case.decl")
    session = Session(entry, {entry: case["source"]})
    eng = session.run().eng
    assert eng is not None and eng.programs is not None and eng.edits is not None
    programs = eng.programs

    def counts():
        return (
            programs.compiled,
            programs.compiled_schemas,
            len(eng.env.registry),
            len(eng.slots_by_key),
            len(eng.reads),
            len(eng.edits.inputs),
        )

    initial = counts()
    for i in range(1, case["iterations"] + 1):
        for q in case["queries"]:
            result = session.evaluate_expr(q["expr"].replace("{i}", str(i)))
            if "error" in q:
                assert result["error"]["code"] == q["error"]
            else:
                assert result["value"] == str(q["factor"] * i + q["offset"])
                assert not result.get("error")
            assert eng.programs is programs and counts() == initial


def test_retained_lifetime() -> None:
    case = json.loads((ROOT / "tests/internal/temporary.json").read_text())["churn"]
    entry = str(ROOT / "tests/internal/temporary_case.decl")
    session = Session(entry, {entry: case["source"]})
    session.apply(
        {"op": "bind", "name": case["root"], "src": {"kind": "inline", "text": case["initial"]}}
    )
    session.run()
    initial = None
    for i in range(case["iterations"]):
        for kind in ("create", "remove"):
            op = {"op": "edit", "kind": kind, "path": f'{case["root"]}["item{i}"]'}
            if kind == "create":
                op["expr"] = case["item"]
            session.apply(op)
            run = session.run()
            eng = run.eng
            assert eng is not None and eng.edits is not None and eng.programs is not None
            assert not run.diags
            assert (
                eng.serialize(eng.env.roots[case["output"]], case["output"])
                == case["present" if kind == "create" else "absent"]
            )
            if kind == "remove":
                counts = (
                    eng.programs.compiled,
                    eng.programs.compiled_schemas,
                    len(eng.env.registry),
                    len(eng.slots_by_key),
                    len(eng.reads),
                    eng.edits.revisions.tracked_queries,
                    len(eng.edits.inputs),
                )
                if initial is None:
                    initial = counts
                assert counts == initial

File: decl-rs/tests/internal/session_test.rs
//! session: the operation log — apply, undo, redo, and a new operation
//! after undo discarding the redo tail.
use crate::common::root;
use decl_lang::semantics::path_str;
use decl_lang::session::{BindSource, EditKind, Mode, Op, Run, Session};
use std::collections::HashMap;
use std::rc::Rc;

#[test]
fn undo_redo() {
    let entry = root().join("tests/repl/documents/main.decl");
    let mut s = Session::new(Some(entry.to_str().unwrap()));
    let bind = |s: &mut Session, text: &str| {
        s.apply(Op::Bind {
            name: "extra".into(),
            src: BindSource::Inline { text: text.into() },
        })
        .unwrap()
    };
    bind(&mut s, "{ \"port\": 1, \"name\": \"x\" }");
    let bound = s.document_text("extra").unwrap();
    assert_eq!(bound, "{\"port\":1,\"name\":\"x\"}");
    assert_eq!(s.undo(1), 1);
    assert!(
        s.document_text("extra").is_err(),
        "undone: the root is invalid"
    );
    assert_eq!(s.redo(1), 1);
    assert_eq!(s.document_text("extra").unwrap(), bound);
    s.undo(1);
    bind(&mut s, "{ \"port\": 2, \"name\": \"y\" }");
    assert_eq!(
        s.redo(1),
        0,
        "a new operation after undo discards the redo tail"
    );
    assert_eq!(
        s.document_text("extra").unwrap(),
        "{\"port\":2,\"name\":\"y\"}"
    );
}

#[test]
fn retained_edits() {
    use super::common::{get, read_json, text};
    use decl_lang::semantics::Value;
    fn field<'a>(v: &'a Value, key: &str) -> &'a str {
        text(get(v, key).unwrap())
    }
    fn list<'a>(v: &'a Value, key: &str) -> &'a [Value] {
        let Some(Value::JArr(a)) = get(v, key) else {
            panic!("{key} is a list")
        };
        a
    }
    fn apply(s: &mut Session, op: &Value) {
        let action = match field(op, "op") {
            "undo" => {
                s.undo(1);
                return;
            }
            "redo" => {
                s.redo(1);
                return;
            }
            "bind" => Op::Bind {
                name: field(op, "name").into(),
                src: BindSource::Inline {
                    text: field(get(op, "src").unwrap(), "text").into(),
                },
            },
            "unbind" => Op::Unbind {
                name: field(op, "name").into(),
            },
            "edit" => Op::Edit {
                kind: match field(op, "kind") {
                    "create" => EditKind::Create,
                    "update" => EditKind::Update,
                    "remove" => EditKind::Remove,
                    _ => panic!("unknown edit"),
                },
                path: field(op, "path").into(),
                expr: get(op, "expr").map(|e| text(e).to_owned()),
            },
            _ => panic!("unknown operation"),
        };
        s.apply(action).unwrap();
    }
    fn snapshot(s: &Session) -> (Run, String) {
        let run = s.run(Mode::Full);
        let values: Vec<_> = run
            .eng
            .as_ref()
            .map(|eng| {
                eng.env
                    .roots_vec()
                    .iter()
                    .map(|(n, v)| (n.clone(), eng.serialize(v, n, false)))
                    .collect()
            })
            .unwrap_or_default();
        let diags = |ds: &[decl_lang::semantics::Diag]| {
            ds.iter().map(|d| d.to_json(None)).collect::<Vec<_>>()
        };
        let text = format!(
            "{:?}",
            (
                values,
                diags(&run.diags),
                diags(&run.load_diags),
                run.checks
                    .iter()
                    .map(|(f, d)| d.to_json(Some(f)))
                    .collect::<Vec<_>>(),
                diags(&run.session_checks)
            )
        );
        (run, text)
    }
    let Value::JArr(cases) =
        read_json(&std::fs::read_to_string(root().join("tests/internal/edits.json")).unwrap())
            .ok()
            .unwrap()
    else {
        panic!("edits.json is a list");
    };
    for case in cases.iter() {
        let label = field(case, "name");
        let entry = root().join("tests/internal/session_case.decl");
        let overlay = HashMap::from([(entry.clone(), field(case, "source").to_owned())]);
        let mut retained = Session::with_overlay(entry.to_str(), Some(&overlay));
        let mut fresh = Session::with_overlay(entry.to_str(), Some(&overlay));
        fresh.full_recompute = true;
        for op in list(case, "initial") {
            apply(&mut retained, op);
            apply(&mut fresh, op);
        }
        let (mut prior, actual) = snapshot(&retained);
        assert!(prior.eng.is_some(), "{label}");
        assert_eq!(actual, snapshot(&fresh).1, "{label} initial");
        for step in list(case, "steps") {
            let eng = prior.eng.as_ref().unwrap();
            let programs = eng.programs.borrow().clone().unwrap();
            let path = get(step, "retained").map(text);
            let inst = eng
                .env
                .registry_snapshot()
                .into_iter()
                .find(|r| Some(path_str(&r.borrow().path, None).as_str()) == path);
            let edits = eng.edits.borrow().clone().unwrap();
            let (verified, equal) = (
                edits.revisions.verified_cutoffs.get(),
                edits.revisions.value_cutoffs.get(),
            );
            apply(&mut retained, get(step, "action").unwrap());
            apply(&mut fresh, get(step, "action").unwrap());
            let (current, actual) = snapshot(&retained);
            assert_eq!(actual, snapshot(&fresh).1, "{label} {step:?}");
            let next = current.eng.as_ref().unwrap();
            assert!(Rc::ptr_eq(
                &programs,
                next.programs.borrow().as_ref().unwrap()
            ));
            if path.is_some() {
                assert!(next
                    .env
                    .registry_snapshot()
                    .iter()
                    .any(|r| Rc::ptr_eq(r, inst.as_ref().unwrap())));
            }
            let number = |key| {
                get(step, key).and_then(|v| {
                    if let Value::Int(n) = v {
                        n.to_string().parse::<usize>().ok()
                    } else {
                        None
                    }
                })
            };
            if let Some(max) = number("max_prepared") {
                assert!(edits.prepared_queries.get() <= max);
            }
            if let Some(max) = number("max_computes") {
                assert!(
                    current.timing.recomputed.unwrap() <= max,
                    "{}",
                    current.timing.recomputed.unwrap()
                );
            }
            if let Some(min) = number("min_verified") {
                assert!(edits.revisions.verified_cutoffs.get() - verified >= min);
            }
            if let Some(min) = number("min_value_cutoffs") {
                assert!(edits.revisions.value_cutoffs.get() - equal >= min);
            }
            prior = current;
        }
    }
}

#[test]
fn temporary_programs() {
    use super::common::{get, read_json, text};
    use decl_lang::semantics::Value;
    let case =
        read_json(&std::fs::read_to_string(root().join("tests/internal/temporary.json")).unwrap())
            .ok()
            .unwrap();
    let number = |row: &Value, key: &str| match get(row, key).unwrap() {
        Value::Int(n) => n.to_string().parse::<usize>().unwrap(),
        _ => panic!("integer expected"),
    };
    let entry = root().join("tests/internal/temporary_case.decl");
    let overlay = HashMap::from([(
        entry.clone(),
        text(get(&case, "source").unwrap()).to_owned(),
    )]);
    let session = Session::with_overlay(entry.to_str(), Some(&overlay));
    let eng = session.run(Mode::Full).eng.unwrap();
    let programs = eng.programs.borrow().clone().unwrap();
    let counts = || {
        (
            programs.compiled.get(),
            programs.compiled_schemas.get(),
            eng.env.registry_snapshot().len(),
            eng.slots_by_key.borrow().len(),
            eng.reads.borrow().len(),
            eng.cached_literals(),
            eng.edits.borrow().as_ref().unwrap().cached_inputs(),
        )
    };
    let initial = counts();
    let Some(Value::JArr(queries)) = get(&case, "queries") else {
        panic!("queries is an array")
    };
    for i in 1..=number(&case, "iterations") {
        for q in queries.iter() {
            let expr = text(get(q, "expr").unwrap()).replace("{i}", &i.to_string());
            let result = session.evaluate_expr(&expr).unwrap();
            if let Some(error) = get(q, "error") {
                assert_eq!(result.error.unwrap().0.as_deref(), Some(text(error)));
            } else {
                assert_eq!(
                    result.value,
                    Some((number(q, "factor") * i + number(q, "offset")).to_string())
                );
                assert!(result.error.is_none());
            }
            assert!(Rc::ptr_eq(
                &programs,
                eng.programs.borrow().as_ref().unwrap()
            ));
            assert_eq!(counts(), initial, "{expr}");
        }
    }
}

#[test]
fn retained_lifetime() {
    use super::common::{get, read_json, text};
    use decl_lang::semantics::Value;
    let cases =
        read_json(&std::fs::read_to_string(root().join("tests/internal/temporary.json")).unwrap())
            .ok()
            .unwrap();
    let case = get(&cases, "churn").unwrap();
    let field = |key| text(get(case, key).unwrap());
    let entry = root().join("tests/internal/temporary_case.decl");
    let overlay = HashMap::from([(entry.clone(), field("source").to_owned())]);
    let mut session = Session::with_overlay(entry.to_str(), Some(&overlay));
    session
        .apply(Op::Bind {
            name: field("root").into(),
            src: BindSource::Inline {
                text: field("initial").into(),
            },
        })
        .unwrap();
    session.run(Mode::Full);
    let Some(Value::Int(iterations)) = get(case, "iterations") else {
        panic!("iterations is an integer")
    };
    let mut initial = None;
    for i in 0..iterations.to_string().parse::<usize>().unwrap() {
        for create in [true, false] {
            session
                .apply(Op::Edit {
                    kind: if create {
                        EditKind::Create
                    } else {
                        EditKind::Remove
                    },
                    path: format!("{}[\"item{i}\"]", field("root")),
                    expr: create.then(|| field("item").to_owned()),
                })
                .unwrap();
            let run = session.run(Mode::Full);
            let eng = run.eng.as_ref().unwrap();
            assert!(run.diags.is_empty(), "{:?}", run.diags);
            assert_eq!(
                eng.serialize(
                    &eng.env.root(field("output")).unwrap(),
                    field("output"),
                    false
                ),
                field(if create { "present" } else { "absent" })
            );
            if !create {
                let programs = eng.programs.borrow().clone().unwrap();
                let edits = eng.edits.borrow().clone().unwrap();
                let counts = (
                    programs.compiled.get(),
                    programs.compiled_schemas.get(),
                    eng.env.registry_snapshot().len(),
                    eng.slots_by_key.borrow().len(),
                    eng.reads.borrow().len(),
                    edits.revisions.tracked_queries(),
                    edits.cached_inputs(),
                    eng.cached_literals(),
                );
                assert_eq!(*initial.get_or_insert(counts), counts);
            }
        }
    }
}
